keps/sig-api-machinery/0014-garbage-collection.md

```diff
@@ -0,0 +1,555 @@
+owner references, but UID is unknown until after a resource is created. Even if
+Kubernetes removes this limitation and allow named based owner references,
+there’s still a resource creation ordering problem: any child resource created
+before its parent resource is prone to be cleaned up by the garbage collector
+unexpectedly.
+
+What's more, a child resource may depend on many other parent resources. It's
+likely for the list of owners of a child resource to be excessively long, and
+hit Kubernetes resource size limit. The resource then cannot be updated anymore
+and becomes useless. 
+
+[UID]: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#uids
+
+## Proposal
+
+The design has three main parts: 
+
+1. **Object-lifecycle-respecting TTL mechanism.** Provide observability before
+   deletion, and address the resource creation ordering problems. See [detailed
+   design of TTL][].
+1. **Garbage collect resources that finish execution.** Delete a resource object
+   when it finishes execution (either terminates or completes). See [detailed
+   design of GC after use (part 1)][].
+1. **Garbage collect resources that are not used by other resources.** Extract
+   usage references of a resource object and delete the object when it is no
+   longer referenced. Garbage collector will count usage references and address
+   the problem of unbounded list of owner references. See [detailed design of GC
+   after use (part 2)][].
+
+This will be coherent with existing owner-reference-based garbage collection
+mechanism. This means that the resource object will stay in existence until all
+garbage collection mechanisms (owner reference, usage references, TTL) agree it
+should be deleted.
+
+[detailed design of TTL]: https://goo.gl/9AUQ3n
+[detailed design of GC after use (part 1)]: https://goo.gl/ws9t3b
+[detailed design of GC after use (part 2)]: https://goo.gl/KeHPiQ
+
+### Generic TTL-based Garbage Collection
+
+First, we need to clarify what TTL means in this context. TTL represents the
+earliest safe time when a resource can be garbage collected. 
+
+Generic TTL-based garbage collection will allow the user to set a TTL on each of
+Kubernetes resources. The TTL starts when a resource is eligible for garbage
+collection and stops when the resource is no longer eligible; the system will
+only garbage collect the resource after the TTL expires. This is useful when the
+user expects a resource to be needed intermittently and/or want to keep the
+resource around for a while before garbage collection kicks in.
+
+Note that we will not use etcd TTL to support this feature, because etcd TTLs
+won’t respect other lifecycle guarantees (e.g. finalizers). This also allows the
+feature to be independent of etcd.
+
+### Clean Up Resources After Use
+
+A resource is no longer *in use* either when it finishes execution (such as a
+terminated Pod or a completed Job), or when it’s not referenced/used by other
```

Comment:
CronJob only needs to see the jobs that haven't completed so that it can react based on its concurrency policy. It doesn't need to see completed jobs. 

The proposed GC will respect owner reference; therefore, Jobs owned by a CronJob won't be deleted by new GC after it completes. It's taken care of by CronJob controller using job history limit.